Binary search benchmark: reduce noise and improve reproducibility.

toziegler opened the pull request:
This PR reduces noise in our binary search benchmark to make potential performance regressions easier to detect.

The benchmark now focuses only on scenarios that matter. Previously, it covered many additional cases, some of which are not relevant in our code, which introduced unnecessary noise.
To further stabilize the results, the benchmark now collects multiple samples, discards the two fastest runs, and reports the third-best result. While this may seem unintuitive, it follows a best practice to mitigate system noise and transient effects as most benchmark distributions are log-normal ( [see](https://lemire.me/blog/2018/01/16/microbenchmarking-calls-for-idealized-conditions/) and  [here](https://github.com/travisdowns/uarch-bench)).

The benchmark also uses a fixed random seed for each benchmark run, improving reproducibility. (The smoke benchmark still uses a random seed.) Finally, an 8 MB allocation was moved from the stack to the heap.

P.S. I want to create a few more benchmarks with @matklad's nice benchmark harness in the next few days.

toziegler committed:
binary search benchmark: reduce signal to noise and make reproducible.

matklad commented on src/lsm/binary_search_benchmark.zig:
```diff
@@ -3,94 +3,123 @@ const assert = std.debug.assert;
+    // Benchmarks require a fixed seed for reproducibility; smoke mode may use a random seed.
+    const seed = bench.parameter(
+        "seed",
+        std.crypto.random.intRangeLessThan(u64, 0, std.math.maxInt(u64)),
```
Tidy complains! It's not exactly on the mark (using crypto.random is ok), but there's still a couple of issues here:

- It feels a bit like layering violation, I'd expect to see `bench.seed()` or `bench.prng`, rather than manually open-coding it. I expect more than a single benchmark to use this!
- This should use `std.testing.random_seed` for reproducibility

I think the call-site (this file) should look like this:

```zig
var prng = stdx.PRNG.from_seed(bench.seed);
```

and then we set bench.seed to testing.random_seed or a fixed constant, depending on mode.

toziegler replied:
This is perfect! I was thinking the same when I wrote the other benchmark today - thanks!

matklad commented on src/lsm/binary_search_benchmark.zig:
```diff
@@ -3,94 +3,123 @@ const assert = std.debug.assert;
     defer arena.deinit();
-
-    const blob = try arena.allocator().alloc(u8, blob_size);
+    const blob = try arena.allocator().alignedAlloc(u8, 64, blob_size);
```
blank like after defer is our house style

matklad commented on src/lsm/binary_search_benchmark.zig:
```diff
@@ -3,94 +3,123 @@ const assert = std.debug.assert;
     search_count: u64,
-    blob: []u8,
+    page_buffer: []u8,
+    allocator: std.mem.Allocator,
```
gpa, or, better, arena (with arena, you won't need defers)

matklad commented on src/lsm/binary_search_benchmark.zig:
```diff
@@ -3,94 +3,123 @@ const assert = std.debug.assert;
-    shuffled_index(prng, value_picker[0..]);
+    const value_picker = try allocator.alloc(usize, layout.values_count);
+    defer allocator.free(value_picker);
+    shuffled_index(prng, value_picker);
```
blank after defer

matklad commented on src/lsm/binary_search_benchmark.zig:
```diff
@@ -99,24 +128,28 @@ fn run_benchmark(
+    std.mem.doNotOptimizeAway(checksum);
+
+    std.sort.block(stdx.Duration, &duration_samples, {}, stdx.Duration.sort.asc);
+    const result = duration_samples[2]; // discard the fastest two, report the 3rd fastest.
```
```suggestion
    const result = duration_samples[2]; // Discard the fastest two, report the 3rd fastest.
```

matklad commented on src/lsm/binary_search_benchmark.zig:
```diff
@@ -99,24 +128,28 @@ fn run_benchmark(
+        duration.* = bench.stop();
+        duration.ns /= search_count;
     }
+    std.mem.doNotOptimizeAway(checksum);
```
bench.report checksum instead of doNotOptimizeAway, to make sure that any change to the behavior of binary search (that is, a bug) is visible to the human.

toziegler replied:
To reduce noise, checksums are accumulated and returned in a single result at the end via `bench.report`. :+1:

toziegler committed:
Update src/lsm/binary_search_benchmark.zig

Co-authored-by: Alex Kladov <aleksey.kladov@gmail.com>

toziegler committed:
binary search benchmark: rename allocator to arena.

toziegler committed:
bench: add `seed` field.
This field returns either a fixed seed for `benchmark` or `std.testing.random_seed` for
`smoke`.

toziegler committed:
binary search benchmark: report checksum.

toziegler commented:
@matklad should we refactor `fn ValueType(comptime layout: Layout) type {`
into `fn ValueType(comptime Key: type, comptime value_size: u32) type`
and move it to `src/testing/fixtures.zig`?
This would make it reusable for the other micro-benchmarks, which I’m working on.

toziegler committed:
binary search benchmark: reorder and defer newline.

matklad commented:
>This would make it reusable for the other micro-benchmarks, which I’m working on.

Your call! The original main goal of `fixtures` was to provide convenient constructs for real objects, rather than entirely fake objects, but, if you want to re-use this, fixtures feels like a natural place! 

Pro tip for expediting lading PRs --- you can always _do_ the change you are unsure about in a separate commit, and then, if you are still unsure, ask the reviewer if that commit should be reverted or not.
